hazelcast/src/main/java/com/hazelcast/query/impl/OrderedIndexStore.java

```diff
@@ -45,7 +45,7 @@
     private volatile Map<Data, QueryableEntry> recordsWithNullValue;
 
     public OrderedIndexStore(IndexCopyBehavior copyOn) {
-        super(copyOn);
+        super(copyOn, true);
```

Comment:
If we are removing the global lock for HD index (which changes the behavior), is there a reason to leave it for other indexes?
Maybe we should introduce a single property that will enable/disable global lock globally?

Reply:
We still leaving it for other indexes. 
* `BitMapIndexStore` is non-concurrent and we still have to keep the lock;
* OrderedIndexStore and `UnorderedIndexStore` as they are written now also non-concurrent. The code below is racy even relying on the concurrent data structures. We have to re-write these operations to make them concurrent. Also, the `compute*` methods of `ConcurrentSkipListMap` which we are using for the `OrderedindexStore` are not atomic according to the JavaDoc - "The function is NOT guaranteed to be applied once atomically."

```
    private class AddFunctor implements IndexFunctor<Comparable, QueryableEntry> {

        @Override
        public Object invoke(Comparable value, QueryableEntry entry) {
            if (value == NULL) {
                return recordsWithNullValue.put(entry.getKeyData(), entry);
            } else {
                Map<Data, QueryableEntry> records = recordMap.get(value);
                if (records == null) {
                    records = new ConcurrentHashMap<>(1, LOAD_FACTOR, 1);
                    recordMap.put(value, records);
                }
                return records.put(entry.getKeyData(), entry);
            }
        }

    }

```

This is the reason why I don't remove the global lock for the on-heap indices.

Reply:
I'm not sure that we have to introduce a single property to enable the global lock, because as it is now we must use it for on-heap indices and if we are able to rewrite the code to make it thread-safe, we just remove them improving throughput.

Reply:
I see, thanks.

Reply:
Looks like we still can relax the locking a bit: take the locks for mutating operations and skip the locking for index queries. Probably there is no that much sense in doing so, the benefits would be visible for the old engine mainly, the new engine would not use the locks for the new iterator-based index query API, if understand things right. WDYT guys?